src/FastRouteRouter.php

```diff
@@ -277,41 +261,50 @@ public function generateUri($name, array $substitutions = [], array $options = [
         }
 
         $route   = $this->routes[$name];
-        $path    = $route->getPath();
         $options = ArrayUtils::merge($route->getOptions(), $options);
 
         if (! empty($options['defaults'])) {
             $substitutions = array_merge($options['defaults'], $substitutions);
         }
 
-        foreach ($substitutions as $key => $value) {
-            $pattern = sprintf(
-                '~%s~x',
-                sprintf(self::VARIABLE_REGEX, preg_quote($key))
-            );
-            $path = preg_replace($pattern, $value, $path);
-        }
+        $routeParser = new RouteParser();
+        $routes = $routeParser->parse($route->getPath());
+
+        // One route pattern can correspond to multiple routes if it has optional parts
+        foreach ($routes as $r) {
+            $path = '';
+            $paramIdx = 0;
+            foreach ($r as $part) {
+                // Fixed segment in the route
+                if (is_string($part)) {
+                    $path .= $part;
+                    continue;
+                }
+
+                // Placeholder in the route
+                if ($paramIdx === count($substitutions)) {
+                    throw new LogicException('Not enough parameters given');
```

Comment:
I think we could make this exception more helpful by indicating how many parameters are required and which were missing.

Each `$part` that is an array appear to be tuples containing the key and a valid regex. If prior to beginning the substitutions we extract all the keys, we'll know the total number of parameters required and what their keys are. `$substitutions` already tells us what we've managed to match, so something along these lines would do the trick (_not tested_):

```php
throw new LogicException(sprintf('Expected parameter values for [%s], but only found [%s]', implode(',', $needed), implode(',', array_keys($substitutions))));
```

... where `$needed` would be that list of all the required parameter names. So instead of `Not enough parameters given`, we'd see `Expected parameter values for [id,foo,bar], but only found [id]` or something like that.

Thoughts @xtreamwayz @weierophinney ?